File: packages/sdk-rn/ios/HybridRovenueNitroSpec.swift
// HybridRovenueNitroSpec.swift — Nitro HybridObject implementation
// forwarding to the M3 Swift `Rovenue` singleton.
//
// Compile target: iOS 13+. Depends on:
//   - NitroModules framework (via react-native-nitro-modules)
//   - Rovenue Swift module (via @rovenue/sdk-swift SPM package)
//
// The Nitrogen tool will normally generate a `HybridRovenueNitroSpecSpec`
// protocol from the .nitro.ts file; for M5 we hand-write the conformance
// against the expected protocol shape. M6's `pnpm nitrogen` step will
// regenerate the protocol; this implementation should remain compatible.
//
// DTO mapping notes (M3 Swift types -> RN-bridge DTOs as declared in
// `src/specs/RovenueNitroSpec.nitro.ts`):
//   * Entitlement: M3 has { id, isActive, productIdentifier, store, expiresIso }
//                  bridge ships { id, active, expiresAt, productId }
//                  (`store` is intentionally dropped on the RN side for now).
//   * ReceiptResult: M3 returns { subscriberId, appUserId, creditBalance }.
//                   M3's `postAppleReceipt` / `postGoogleReceipt` only resolve
//                   on success and the docstring guarantees entitlements +
//                   credits caches were refreshed on success. We therefore
//                   synthesize { ok: true, entitlementsRefreshed: true,
//                                creditsRefreshed: true } when the call
//                   returns; failures propagate as thrown errors.
//   * `Rovenue.shared.currentUser()` is `async` (non-throwing) — we expose it
//     as `async throws` to match the Nitro spec but never actually throw.
//   * `Rovenue.shared.entitlement(_:)`, `entitlementsAll()`, `creditBalance()`
//     are `async` (non-throwing) — same treatment.
//   * Most M3 methods use unlabeled first params (`identify(_:)`,
//     `entitlement(_:)`, `consumeCredits(_:description:)`,
//     `postAppleReceipt(_:productId:)`, `postGoogleReceipt(_:productId:)`).

import Foundation
import NitroModules
import Rovenue  // Swift M3 package

final class HybridRovenueNitroSpec: HybridObject {

    // -------- Lifecycle --------
    func configure(apiKey: String, baseUrl: String, debug: Bool) throws {
        try Rovenue.configure(apiKey: apiKey, baseUrl: baseUrl, debug: debug)
    }

    func shutdown() {
        Rovenue.shared.shutdown()
    }

    func setForeground(foreground: Bool) {
        Rovenue.shared.setForeground(foreground)
    }

    func getVersion() -> String {
        return Rovenue.shared.version
    }

    // -------- Identity --------
    func currentUser() async throws -> [String: Any?] {
        let u = await Rovenue.shared.currentUser()
        return [
            "anonId": u.anonId,
            "knownUserId": u.knownUserId as Any?,
        ]
    }

    func identify(knownUserId: String) async throws {
        try await Rovenue.shared.identify(knownUserId)
    }

    // -------- Entitlements --------
    func entitlement(id: String) async throws -> [String: Any?]? {
        guard let e = await Rovenue.shared.entitlement(id) else { return nil }
        return Self.dtoFromEntitlement(e)
    }

    func entitlementsAll() async throws -> [[String: Any?]] {
        let all = await Rovenue.shared.entitlementsAll()
        return all.map(Self.dtoFromEntitlement)
    }

    func refreshEntitlements() async throws {
        try await Rovenue.shared.refreshEntitlements()
    }

    // -------- Credits --------
    func creditBalance() async throws -> Double {
        let b = await Rovenue.shared.creditBalance()
        return Double(b)  // Nitro JS Number is Double; Int64 → Double is lossless up to 2^53
    }

    func refreshCredits() async throws {
        try await Rovenue.shared.refreshCredits()
    }

    func consumeCredits(amount: Double, description: String?) async throws -> Double {
        let b = try await Rovenue.shared.consumeCredits(
            Int64(amount),
            description: description
        )
        return Double(b)
    }

    // -------- Receipts --------
    func postAppleReceipt(jws: String, productId: String) async throws -> [String: Any?] {
        _ = try await Rovenue.shared.postAppleReceipt(jws, productId: productId)
        // M3 only returns on success and guarantees both caches were refreshed.
        return [
            "ok": true,
            "entitlementsRefreshed": true,
            "creditsRefreshed": true,
        ]
    }

    func postGoogleReceipt(receipt: String, productId: String) async throws -> [String: Any?] {
        _ = try await Rovenue.shared.postGoogleReceipt(receipt, productId: productId)
        return [
            "ok": true,
            "entitlementsRefreshed": true,
            "creditsRefreshed": true,
        ]
    }

    // -------- Observer --------
    func addChangeListener(cb: @escaping (String) -> Void) -> () -> Void {
        let task = Task {
            for await event in Rovenue.shared.changes {
                cb(Self.eventName(event))
            }
        }
        return { task.cancel() }
    }

    // -------- Helpers --------
    private static func dtoFromEntitlement(_ e: Entitlement) -> [String: Any?] {
        return [
            "id": e.id,
            "active": e.isActive,
            "expiresAt": e.expiresIso as Any?,
            "productId": e.productIdentifier as Any?,
        ]
    }

    private static func eventName(_ event: ChangeEvent) -> String {
        switch event {
        case .entitlementsChanged:    return "ENTITLEMENTS_CHANGED"
        case .identityChanged:        return "IDENTITY_CHANGED"
        case .creditBalanceChanged:   return "CREDIT_BALANCE_CHANGED"
        }
    }
}

File: packages/sdk-rn/android/src/main/java/dev/rovenue/sdkrn/HybridRovenueNitroSpec.kt
// HybridRovenueNitroSpec.kt — Nitro HybridObject implementation
// forwarding to the M4 Kotlin `Rovenue` singleton.
//
// Compile target: Android 24+ (JVM 17 baseline running on ART).
// Depends on:
//   - Nitro Modules Android runtime
//   - :sdk-kotlin (M4 façade Gradle module)
//
// M5 SCOPE NOTE
// -------------
// This file is checked in source-only; the actual Android library
// build is performed by the consuming RN app's autolinking step in
// M6. Nitrogen is also wired up in M6 and will normally generate a
// `HybridRovenueNitroSpecSpec` base class from the .nitro.ts spec —
// we hand-write the conformance against the expected shape here.
//
// M4 SURFACE STATUS
// -----------------
// At the time of writing, `packages/sdk-kotlin/src/main/kotlin/dev/
// rovenue/sdk/Rovenue.kt` only exposes `Rovenue.version` (a static
// `val`). The richer surface used below — `Rovenue.configure(...)`,
// `Rovenue.shared.identify(...)`, etc. — is the *expected* M4 façade
// shape that mirrors the M3 Swift `Rovenue` API. M4 is expected to
// grow to match (companion object for static `configure`, `shared`
// instance for the rest) when the Kotlin façade is fleshed out.
// Until then this file will NOT compile against M4; revisiting in M6
// is required (same as the iOS Swift spec).
//
// DTO mapping notes (UniFFI-generated Kotlin types -> RN-bridge DTOs
// as declared in `src/specs/RovenueNitroSpec.nitro.ts`):
//   * Entitlement: UDL fields `is_active`, `product_identifier`,
//                  `expires_iso`, `store` → Kotlin camelCase
//                  `isActive`, `productIdentifier`, `expiresIso`,
//                  `store`. RN ships `{ id, active, expiresAt,
//                  productId }` — `store` is intentionally dropped.
//   * ReceiptResult: UDL exposes `{subscriberId, appUserId,
//                    creditBalance}`. The suspend functions only
//                    return on success and (per spec) guarantee
//                    entitlements + credits caches were refreshed.
//                    We therefore synthesise `{ok: true,
//                    entitlementsRefreshed: true,
//                    creditsRefreshed: true}` when the call returns;
//                    failures propagate as thrown exceptions.
//   * `Rovenue.shared.changes` is the `SharedFlow<ChangeEvent>` from M4.
//   * `ChangeEvent.name` returns the SCREAMING_SNAKE enum constant name,
//     matching the iOS switch's output verbatim
//     (ENTITLEMENTS_CHANGED, IDENTITY_CHANGED, CREDIT_BALANCE_CHANGED).
//   * Kotlin `suspend` functions can throw — adjust if M4 marks any
//     of these non-throwing.

package dev.rovenue.sdkrn

import com.margelo.nitro.HybridObject
import dev.rovenue.sdk.Rovenue
import dev.rovenue.sdk.generated.ChangeEvent
import dev.rovenue.sdk.generated.Entitlement
import dev.rovenue.sdk.generated.ReceiptResult
import dev.rovenue.sdk.generated.User
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch

class HybridRovenueNitroSpec : HybridObject() {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    // -------- Lifecycle --------
    fun configure(apiKey: String, baseUrl: String, debug: Boolean) {
        Rovenue.configure(apiKey, baseUrl, debug)
    }

    fun shutdown() {
        Rovenue.shared.shutdown()
    }

    fun setForeground(foreground: Boolean) {
        Rovenue.shared.setForeground(foreground)
    }

    fun getVersion(): String = Rovenue.shared.version

    // -------- Identity --------
    suspend fun currentUser(): Map<String, Any?> {
        val u: User = Rovenue.shared.currentUser()
        return mapOf(
            "anonId" to u.anonId,
            "knownUserId" to u.knownUserId,
        )
    }

    suspend fun identify(knownUserId: String) {
        Rovenue.shared.identify(knownUserId)
    }

    // -------- Entitlements --------
    suspend fun entitlement(id: String): Map<String, Any?>? {
        val e: Entitlement = Rovenue.shared.entitlement(id) ?: return null
        return dtoFromEntitlement(e)
    }

    suspend fun entitlementsAll(): List<Map<String, Any?>> {
        return Rovenue.shared.entitlementsAll().map(::dtoFromEntitlement)
    }

    suspend fun refreshEntitlements() {
        Rovenue.shared.refreshEntitlements()
    }

    // -------- Credits --------
    suspend fun creditBalance(): Double {
        // M4 returns Long (Int64 from UDL `i64`); Nitro JS Number is
        // Double — lossless up to 2^53.
        return Rovenue.shared.creditBalance().toDouble()
    }

    suspend fun refreshCredits() {
        Rovenue.shared.refreshCredits()
    }

    suspend fun consumeCredits(amount: Double, description: String?): Double {
        return Rovenue.shared.consumeCredits(amount.toLong(), description).toDouble()
    }

    // -------- Receipts --------
    suspend fun postAppleReceipt(jws: String, productId: String): Map<String, Any?> {
        // Discard the M4 ReceiptResult — only the success/throw
        // distinction is meaningful at the RN bridge layer.
        Rovenue.shared.postAppleReceipt(jws, productId)
        return mapOf(
            "ok" to true,
            "entitlementsRefreshed" to true,
            "creditsRefreshed" to true,
        )
    }

    suspend fun postGoogleReceipt(receipt: String, productId: String): Map<String, Any?> {
        Rovenue.shared.postGoogleReceipt(receipt, productId)
        return mapOf(
            "ok" to true,
            "entitlementsRefreshed" to true,
            "creditsRefreshed" to true,
        )
    }

    // -------- Observer --------
    fun addChangeListener(cb: (String) -> Unit): () -> Unit {
        val job: Job = scope.launch {
            Rovenue.shared.changes.collect { event ->
                cb(eventName(event))
            }
        }
        return { job.cancel() }
    }

    // -------- Helpers --------
    private fun dtoFromEntitlement(e: Entitlement): Map<String, Any?> {
        return mapOf(
            "id" to e.id,
            "active" to e.isActive,
            "expiresAt" to e.expiresIso,
            "productId" to e.productIdentifier,
        )
    }

    private fun eventName(event: ChangeEvent): String = event.name
}
